File: java/src/P029/DivideTwoIntegers/DivideTwoIntegers.java
package P029.DivideTwoIntegers;

public class DivideTwoIntegers {
    public int divide(int dividend, int divisor) {
        if (dividend == Integer.MIN_VALUE) {
            if (divisor == -1) return Integer.MAX_VALUE;
            else if (divisor == 1) return dividend;
        }

        int sign = (dividend < 0) ^ (divisor < 0) ? -1 : 1;
        long dvd = Math.abs((long)dividend);
        long dvs = Math.abs((long)divisor);

        int res = 0;
        while (dvd >= dvs) {
            long tmp = dvs;
            long multiple = 1;
            while (dvd >= (tmp << 1)) {
                tmp <<= 1;
                multiple <<= 1;
            }
            res += multiple;
            dvd -= tmp;
        }
        return sign == 1 ? res : -res;
    }
}

File: python/P029_DivideTwoIntegers/DivideTwoIntegers.py
# -*- coding: utf-8 -*-
class Solution(object):
    def divide(self, dividend, divisor):
        MIN, MAX = -2147483648, 2147483647
        positive = (dividend < 0) is (divisor < 0)
        dividend, divisor = abs(dividend), abs(divisor)
        res = 0
        while dividend >= divisor:
            tmp, i = divisor, 1
            while dividend >= tmp:
                dividend -= tmp
                res += i
                tmp, i = tmp << 1, i << 1
        if not positive:
            res = -res
        return min(max(MIN, res), MAX)
